src/Framework/Blocks/BlockCollection.php

```diff
@@ -79,24 +79,117 @@ public function getBlocks(): array
-                return $this->findByName($name, $block->innerBlocks->blocks);
+        foreach ($blockCollection->blocks as $block) {
+            if ($block->name === $blockName) {
+                $count++;
+
+                if ($count === $blockIndex + 1) {
+                    if ($return === 'self') {
+                        return $block;
+                    } elseif ($return === 'parent') {
+                        return $blockCollection;
+                    }
+                }
+            } elseif ($block->innerBlocks) {
+                $result = $this->findByNameRecursive($blockName, $blockIndex, $return, $block->innerBlocks, $count);
+                if ($result) {
+                    return $result;
+                }
             }
         }
 
         return null;
     }
+
+    /**
+     * @unreleased
+     */
+    public function insertBefore(string $blockName, BlockModel $block, int $blockIndex = 0): BlockCollection
+    {
+        $blockCollection = $this->findByName($blockName, $blockIndex, 'parent');
+
+        if (!$blockCollection) {
+            return $this;
+        }
+
+        $innerBlocks = $blockCollection->blocks;
+        $blockIndex = array_search($blockName, array_column($innerBlocks, 'name'));
+        array_splice($innerBlocks, $blockIndex, 0, [$block]);
+        $blockCollection->blocks = $innerBlocks;
+
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function insertAfter(string $blockName, BlockModel $block, int $blockIndex = 0): BlockCollection
+    {
+        $blockCollection = $this->findByName($blockName, $blockIndex, 'parent');
+
+        if (!$blockCollection) {
+            return $this;
+        }
+
+        $innerBlocks = $blockCollection->blocks;
+        $blockIndex = array_search($blockName, array_column($innerBlocks, 'name'));
+        array_splice($innerBlocks, $blockIndex + 1, 0, [$block]);
+        $blockCollection->blocks = $innerBlocks;
+
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function prepend(string $blockName, BlockModel $block, int $blockIndex = 0): BlockCollection
+    {
+        $blockCollection = $this->findByName($blockName, $blockIndex)->innerBlocks;
+
+        if (!$blockCollection) {
+            return $this;
+        }
+
+        $innerBlocks = $blockCollection->blocks;
+        array_unshift($innerBlocks, $block);
+        $blockCollection->blocks = $innerBlocks;
+
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function append(string $blockName, BlockModel $block, int $blockIndex = 0): BlockCollection
+    {
+        $blockCollection = $this->findByName($blockName, $blockIndex)->innerBlocks;
+
+        if (!$blockCollection) {
+            return $this;
+        }
+
+        $innerBlocks = $blockCollection->blocks;
+        $innerBlocks[] = $block;
+        $blockCollection->blocks = $innerBlocks;
+
+        return $this;
+    }
 }
```

Comment:
Can you please update the `TestBlockCollection` unit test to include this new api?  I want us to be confident these work properly since they are going to be immediately used in add-ons.